File: Backend/FloraX/FloraX/src/main/java/com/example/FloraX/Service/UserDashboardService.java
package com.example.FloraX.Service;

import com.example.FloraX.Dto.*;

import java.util.List;

/**
 * UserDashboardService — Complete Service Interface
 *
 * Provides all operations needed to power the user dashboard:
 * - Full dashboard snapshot
 * - Summary statistics
 * - Garden / Zone details
 * - Sensor & sensor readings
 * - Irrigation logs (daily / weekly / monthly)
 * - Alerts (active, resolved, by type)
 * - Water tank status
 * - Valve status
 *
 * All methods accept the authenticated user's email (extracted from JWT).
 */
public interface UserDashboardService {

    // ══════════════════════════════════════════════════════════════════════════
    // 1. FULL DASHBOARD
    // ══════════════════════════════════════════════════════════════════════════

    /**
     * Returns the complete dashboard response for the authenticated user.
     * Includes summary stats, all gardens with zones, sensors, alerts,
     * recent irrigation logs, and water tank status.
     */
    UserDashboardResponse getDashboard(String email);

    // ══════════════════════════════════════════════════════════════════════════
    // 2. SUMMARY STATISTICS
    // ══════════════════════════════════════════════════════════════════════════

    /**
     * Returns aggregated summary stats for all gardens belonging to the user:
     * total gardens, zones, sensors, active/faulty sensors, valves,
     * water usage today/week/month, alert counts by type, tank levels.
     */
    DashboardSummaryDTO getDashboardSummary(String email);

    // ══════════════════════════════════════════════════════════════════════════
    // 3. GARDEN OPERATIONS
    // ══════════════════════════════════════════════════════════════════════════

    /**
     * Returns all gardens belonging to the user with their zones,
     * recent alerts, and irrigation logs.
     */
    List<GardenDTO> getAllGardens(String email);

    /**
     * Returns a single garden by ID.
     * Throws RuntimeException if the garden doesn't belong to the user.
     */
    GardenDTO getGardenById(String email, Long gardenId);

    // ══════════════════════════════════════════════════════════════════════════
    // 4. ZONE OPERATIONS
    // ══════════════════════════════════════════════════════════════════════════

    /**
     * Returns all zones across all gardens of the user,
     * each with latest moisture reading and irrigation status.
     */
    List<ZoneDTO> getAllZones(String email);

    /**
     * Returns all zones belonging to the specified garden,
     * each with the latest sensor reading and status.
     */
    List<ZoneDTO> getZonesByGarden(String email, Long gardenId);

    /**
     * Returns a single zone by ID with its sensors, latest readings,
     * and last irrigation log.
     */
    ZoneDTO getZoneById(String email, Long zoneId);

    /**
     * Returns all zones currently in ALERT status (moisture below threshold).
     */
    List<ZoneDTO> getAlertZones(String email);

    /**
     * Returns all zones that are actively being irrigated right now.
     */
    List<ZoneDTO> getActiveIrrigationZones(String email);

    // ══════════════════════════════════════════════════════════════════════════
    // 5. SENSOR OPERATIONS
    // ══════════════════════════════════════════════════════════════════════════

    /**
     * Returns all sensors across all zones of the user,
     * each with their latest reading.
     */
    List<SensorDTO> getAllSensors(String email);

    /**
     * Returns all sensors belonging to the specified zone,
     * with latest readings.
     */
    List<SensorDTO> getSensorsByZone(String email, Long zoneId);

    /**
     * Returns only sensors with FAULTY or INACTIVE status.
     */
    List<SensorDTO> getFaultySensors(String email);

    // ══════════════════════════════════════════════════════════════════════════
    // 6. IRRIGATION LOG OPERATIONS
    // ══════════════════════════════════════════════════════════════════════════

    /**
     * Returns all irrigation logs from today across all gardens of the user.
     */
    List<IrrigationLogDTO> getTodayIrrigationLogs(String email);

    /**
     * Returns irrigation logs from the past 7 days.
     */
    List<IrrigationLogDTO> getWeeklyIrrigationLogs(String email);

    /**
     * Returns irrigation logs from the past 30 days.
     */
    List<IrrigationLogDTO> getMonthlyIrrigationLogs(String email);

    /**
     * Returns the last N irrigation logs across all user gardens.
     */
    List<IrrigationLogDTO> getRecentIrrigationLogs(String email, int limit);

    /**
     * Returns irrigation logs for a specific zone.
     */
    List<IrrigationLogDTO> getIrrigationLogsByZone(String email, Long zoneId, int limit);

    /**
     * Returns the total water volume used today (litres) across all gardens.
     */
    Double getTotalWaterUsedToday(String email);

    /**
     * Returns the total water volume used this week (litres).
     */
    Double getTotalWaterUsedThisWeek(String email);

    /**
     * Returns the total water volume used this month (litres).
     */
    Double getTotalWaterUsedThisMonth(String email);

    // ══════════════════════════════════════════════════════════════════════════
    // 7. ALERT OPERATIONS
    // ══════════════════════════════════════════════════════════════════════════

    /**
     * Returns all ACTIVE alerts across all gardens of the user.
     */
    List<AlertDTO> getActiveAlerts(String email);

    /**
     * Returns alerts resolved today.
     */
    List<AlertDTO> getResolvedAlertsToday(String email);

    /**
     * Returns the last N alerts (any status) across all gardens.
     */
    List<AlertDTO> getRecentAlerts(String email, int limit);

    /**
     * Returns alerts for a specific garden.
     */
    List<AlertDTO> getAlertsByGarden(String email, Long gardenId);

    /**
     * Returns the count of active alerts grouped by alert type.
     * Key = AlertType name, Value = count.
     */
    java.util.Map<String, Long> getAlertCountByType(String email);

    /**
     * Marks an alert as RESOLVED.
     * Throws exception if the alert doesn't belong to the user.
     */
    void resolveAlert(String email, Long alertId);

    // ══════════════════════════════════════════════════════════════════════════
    // 8. WATER TANK OPERATIONS
    // ══════════════════════════════════════════════════════════════════════════

    /**
     * Returns all water tanks across all gardens of the user,
     * with their current fill level and status.
     */
    List<WaterTankDTO> getAllWaterTanks(String email);

    /**
     * Returns all water tanks for a specific garden.
     */
    List<WaterTankDTO> getWaterTanksByGarden(String email, Long gardenId);

    /**
     * Returns tanks that are LOW or CRITICAL on water level.
     */
    List<WaterTankDTO> getLowWaterTanks(String email);

    // ══════════════════════════════════════════════════════════════════════════
    // 9. VALVE OPERATIONS
    // ══════════════════════════════════════════════════════════════════════════

    /**
     * Returns all valves across all zones of the user,
     * with current status (OPEN, CLOSED, FAULTY).
     */
    List<ValveDTO> getAllValves(String email);

    /**
     * Returns all valves in a specific zone.
     */
    List<ValveDTO> getValvesByZone(String email, Long zoneId);

    /**
     * Returns all valves that are currently OPEN (actively irrigating).
     */
    List<ValveDTO> getOpenValves(String email);
}

File: Frontend/reactapp/src/api/userDashboardService.js
/**
 * userDashboardService.js
 *
 * Frontend API service for the FloraX User Dashboard.
 * Maps 1-to-1 with UserDashboardController endpoints.
 *
 * Base URL  : http://localhost:8081/florax/api
 * Auth      : JWT Bearer token auto-attached via axios interceptor
 * Prefix    : /dashboard
 *
 * Usage example:
 *   import { getMyDashboard, getActiveAlerts } from '../api/userDashboardService';
 *   const data = await getMyDashboard();
 */

import api from "./axios";

const BASE = "/dashboard";

// ══════════════════════════════════════════════════════════════════════════════
// UTILITY — centralised error extractor
// ══════════════════════════════════════════════════════════════════════════════
const extractError = (error) => {
    const msg =
        error?.response?.data?.message ||
        error?.response?.data ||
        error?.message ||
        "An unexpected error occurred";
    throw new Error(msg);
};

// ══════════════════════════════════════════════════════════════════════════════
// 1. FULL DASHBOARD
// ══════════════════════════════════════════════════════════════════════════════

/**
 * GET /dashboard/me
 * Returns the complete dashboard snapshot for the logged-in user.
 * Includes summary stats, all gardens, zones, sensors, alerts, and irrigation logs.
 *
 * @returns {Promise<UserDashboardResponse>}
 */
export const getMyDashboard = async () => {
    try {
        const res = await api.get(`${BASE}/me`);
        return res.data;
    } catch (error) {
        extractError(error);
    }
};

// ══════════════════════════════════════════════════════════════════════════════
// 2. SUMMARY STATISTICS
// ══════════════════════════════════════════════════════════════════════════════

/**
 * GET /dashboard/summary
 * Returns aggregated stats: total gardens, zones, sensors, valves,
 * water usage (today / week / month), alert counts by type, and tank levels.
 *
 * @returns {Promise<DashboardSummaryDTO>}
 */
export const getDashboardSummary = async () => {
    try {
        const res = await api.get(`${BASE}/summary`);
        return res.data;
    } catch (error) {
        extractError(error);
    }
};

// ══════════════════════════════════════════════════════════════════════════════
// 3. GARDEN OPERATIONS
// ══════════════════════════════════════════════════════════════════════════════

/**
 * GET /dashboard/gardens
 * Returns all gardens with their zones, recent alerts, and irrigation logs.
 *
 * @returns {Promise<GardenDTO[]>}
 */
export const getAllGardens = async () => {
    try {
        const res = await api.get(`${BASE}/gardens`);
        return res.data;
    } catch (error) {
        extractError(error);
    }
};

/**
 * GET /dashboard/gardens/{gardenId}
 * Returns a single garden by ID with full zone and alert detail.
 *
 * @param {number} gardenId
 * @returns {Promise<GardenDTO>}
 */
export const getGardenById = async (gardenId) => {
    try {
        const res = await api.get(`${BASE}/gardens/${gardenId}`);
        return res.data;
    } catch (error) {
        extractError(error);
    }
};

// ══════════════════════════════════════════════════════════════════════════════
// 4. ZONE OPERATIONS
// ══════════════════════════════════════════════════════════════════════════════

/**
 * GET /dashboard/zones
 * Returns all zones across all gardens with live moisture and irrigation status.
 *
 * @returns {Promise<ZoneDTO[]>}
 */
export const getAllZones = async () => {
    try {
        const res = await api.get(`${BASE}/zones`);
        return res.data;
    } catch (error) {
        extractError(error);
    }
};

/**
 * GET /dashboard/gardens/{gardenId}/zones
 * Returns all zones in a specific garden.
 *
 * @param {number} gardenId
 * @returns {Promise<ZoneDTO[]>}
 */
export const getZonesByGarden = async (gardenId) => {
    try {
        const res = await api.get(`${BASE}/gardens/${gardenId}/zones`);
        return res.data;
    } catch (error) {
        extractError(error);
    }
};

/**
 * GET /dashboard/zones/{zoneId}
 * Returns a single zone by ID with sensors, latest reading, and last irrigation.
 *
 * @param {number} zoneId
 * @returns {Promise<ZoneDTO>}
 */
export const getZoneById = async (zoneId) => {
    try {
        const res = await api.get(`${BASE}/zones/${zoneId}`);
        return res.data;
    } catch (error) {
        extractError(error);
    }
};

/**
 * GET /dashboard/zones/alert
 * Returns all zones currently below their moisture threshold (ALERT status).
 *
 * @returns {Promise<ZoneDTO[]>}
 */
export const getAlertZones = async () => {
    try {
        const res = await api.get(`${BASE}/zones/alert`);
        return res.data;
    } catch (error) {
        extractError(error);
    }
};

/**
 * GET /dashboard/zones/active
 * Returns all zones that are currently being irrigated (ACTIVE status).
 *
 * @returns {Promise<ZoneDTO[]>}
 */
export const getActiveIrrigationZones = async () => {
    try {
        const res = await api.get(`${BASE}/zones/active`);
        return res.data;
    } catch (error) {
        extractError(error);
    }
};

// ══════════════════════════════════════════════════════════════════════════════
// 5. SENSOR OPERATIONS
// ══════════════════════════════════════════════════════════════════════════════

/**
 * GET /dashboard/sensors
 * Returns all sensors across all zones with their latest reading.
 *
 * @returns {Promise<SensorDTO[]>}
 */
export const getAllSensors = async () => {
    try {
        const res = await api.get(`${BASE}/sensors`);
        return res.data;
    } catch (error) {
        extractError(error);
    }
};

/**
 * GET /dashboard/zones/{zoneId}/sensors
 * Returns all sensors for a specific zone with latest readings.
 *
 * @param {number} zoneId
 * @returns {Promise<SensorDTO[]>}
 */
export const getSensorsByZone = async (zoneId) => {
    try {
        const res = await api.get(`${BASE}/zones/${zoneId}/sensors`);
        return res.data;
    } catch (error) {
        extractError(error);
    }
};

/**
 * GET /dashboard/sensors/faulty
 * Returns all sensors with FAULTY or INACTIVE status.
 *
 * @returns {Promise<SensorDTO[]>}
 */
export const getFaultySensors = async () => {
    try {
        const res = await api.get(`${BASE}/sensors/faulty`);
        return res.data;
    } catch (error) {
        extractError(error);
    }
};

// ══════════════════════════════════════════════════════════════════════════════
// 6. IRRIGATION LOG OPERATIONS
// ══════════════════════════════════════════════════════════════════════════════

/**
 * GET /dashboard/irrigation/today
 * Returns all irrigation logs from today across all gardens.
 *
 * @returns {Promise<IrrigationLogDTO[]>}
 */
export const getTodayIrrigationLogs = async () => {
    try {
        const res = await api.get(`${BASE}/irrigation/today`);
        return res.data;
    } catch (error) {
        extractError(error);
    }
};

/**
 * GET /dashboard/irrigation/weekly
 * Returns irrigation logs from the past 7 days.
 *
 * @returns {Promise<IrrigationLogDTO[]>}
 */
export const getWeeklyIrrigationLogs = async () => {
    try {
        const res = await api.get(`${BASE}/irrigation/weekly`);
        return res.data;
    } catch (error) {
        extractError(error);
    }
};

/**
 * GET /dashboard/irrigation/monthly
 * Returns irrigation logs from the past 30 days.
 *
 * @returns {Promise<IrrigationLogDTO[]>}
 */
export const getMonthlyIrrigationLogs = async () => {
    try {
        const res = await api.get(`${BASE}/irrigation/monthly`);
        return res.data;
    } catch (error) {
        extractError(error);
    }
};

/**
 * GET /dashboard/irrigation/recent?limit=10
 * Returns the last N irrigation logs across all gardens.
 *
 * @param {number} [limit=10]
 * @returns {Promise<IrrigationLogDTO[]>}
 */
export const getRecentIrrigationLogs = async (limit = 10) => {
    try {
        const res = await api.get(`${BASE}/irrigation/recent`, { params: { limit } });
        return res.data;
    } catch (error) {
        extractError(error);
    }
};

/**
 * GET /dashboard/zones/{zoneId}/irrigation?limit=10
 * Returns irrigation logs for a specific zone.
 *
 * @param {number} zoneId
 * @param {number} [limit=10]
 * @returns {Promise<IrrigationLogDTO[]>}
 */
export const getIrrigationLogsByZone = async (zoneId, limit = 10) => {
    try {
        const res = await api.get(`${BASE}/zones/${zoneId}/irrigation`, { params: { limit } });
        return res.data;
    } catch (error) {
        extractError(error);
    }
};

/**
 * GET /dashboard/water-usage/today
 * Returns total water volume used today (litres).
 *
 * @returns {Promise<number>}
 */
export const getWaterUsageToday = async () => {
    try {
        const res = await api.get(`${BASE}/water-usage/today`);
        return res.data;
    } catch (error) {
        extractError(error);
    }
};

/**
 * GET /dashboard/water-usage/weekly
 * Returns total water volume used this week (litres).
 *
 * @returns {Promise<number>}
 */
export const getWaterUsageWeekly = async () => {
    try {
        const res = await api.get(`${BASE}/water-usage/weekly`);
        return res.data;
    } catch (error) {
        extractError(error);
    }
};

/**
 * GET /dashboard/water-usage/monthly
 * Returns total water volume used this month (litres).
 *
 * @returns {Promise<number>}
 */
export const getWaterUsageMonthly = async () => {
    try {
        const res = await api.get(`${BASE}/water-usage/monthly`);
        return res.data;
    } catch (error) {
        extractError(error);
    }
};

// ══════════════════════════════════════════════════════════════════════════════
// 7. ALERT OPERATIONS
// ══════════════════════════════════════════════════════════════════════════════

/**
 * GET /dashboard/alerts
 * Returns all ACTIVE alerts across all gardens.
 *
 * @returns {Promise<AlertDTO[]>}
 */
export const getActiveAlerts = async () => {
    try {
        const res = await api.get(`${BASE}/alerts`);
        return res.data;
    } catch (error) {
        extractError(error);
    }
};

/**
 * GET /dashboard/alerts/resolved-today
 * Returns alerts that were resolved today.
 *
 * @returns {Promise<AlertDTO[]>}
 */
export const getResolvedAlertsToday = async () => {
    try {
        const res = await api.get(`${BASE}/alerts/resolved-today`);
        return res.data;
    } catch (error) {
        extractError(error);
    }
};

/**
 * GET /dashboard/alerts/recent?limit=10
 * Returns the last N alerts (any status), sorted newest first.
 *
 * @param {number} [limit=10]
 * @returns {Promise<AlertDTO[]>}
 */
export const getRecentAlerts = async (limit = 10) => {
    try {
        const res = await api.get(`${BASE}/alerts/recent`, { params: { limit } });
        return res.data;
    } catch (error) {
        extractError(error);
    }
};

/**
 * GET /dashboard/gardens/{gardenId}/alerts
 * Returns alerts for a specific garden.
 *
 * @param {number} gardenId
 * @returns {Promise<AlertDTO[]>}
 */
export const getAlertsByGarden = async (gardenId) => {
    try {
        const res = await api.get(`${BASE}/gardens/${gardenId}/alerts`);
        return res.data;
    } catch (error) {
        extractError(error);
    }
};

/**
 * GET /dashboard/alerts/count-by-type
 * Returns alert counts grouped by type.
 * Example: { "LOW_WATER": 2, "DRY_SOIL": 1, "SENSOR_FAULT": 0 }
 *
 * @returns {Promise<Record<string, number>>}
 */
export const getAlertCountByType = async () => {
    try {
        const res = await api.get(`${BASE}/alerts/count-by-type`);
        return res.data;
    } catch (error) {
        extractError(error);
    }
};

/**
 * PUT /dashboard/alerts/{alertId}/resolve
 * Marks an alert as RESOLVED. Returns 204 No Content on success.
 *
 * @param {number} alertId
 * @returns {Promise<void>}
 */
export const resolveAlert = async (alertId) => {
    try {
        await api.put(`${BASE}/alerts/${alertId}/resolve`);
    } catch (error) {
        extractError(error);
    }
};

// ══════════════════════════════════════════════════════════════════════════════
// 8. WATER TANK OPERATIONS
// ══════════════════════════════════════════════════════════════════════════════

/**
 * GET /dashboard/tanks
 * Returns all water tanks with current fill level and percentage.
 *
 * @returns {Promise<WaterTankDTO[]>}
 */
export const getAllWaterTanks = async () => {
    try {
        const res = await api.get(`${BASE}/tanks`);
        return res.data;
    } catch (error) {
        extractError(error);
    }
};

/**
 * GET /dashboard/gardens/{gardenId}/tanks
 * Returns water tanks for a specific garden.
 *
 * @param {number} gardenId
 * @returns {Promise<WaterTankDTO[]>}
 */
export const getWaterTanksByGarden = async (gardenId) => {
    try {
        const res = await api.get(`${BASE}/gardens/${gardenId}/tanks`);
        return res.data;
    } catch (error) {
        extractError(error);
    }
};

/**
 * GET /dashboard/tanks/low
 * Returns tanks with LOW or EMPTY status — needs refilling.
 *
 * @returns {Promise<WaterTankDTO[]>}
 */
export const getLowWaterTanks = async () => {
    try {
        const res = await api.get(`${BASE}/tanks/low`);
        return res.data;
    } catch (error) {
        extractError(error);
    }
};

// ══════════════════════════════════════════════════════════════════════════════
// 9. VALVE OPERATIONS
// ══════════════════════════════════════════════════════════════════════════════

/**
 * GET /dashboard/valves
 * Returns all valves with their current status and last activation time.
 *
 * @returns {Promise<ValveDTO[]>}
 */
export const getAllValves = async () => {
    try {
        const res = await api.get(`${BASE}/valves`);
        return res.data;
    } catch (error) {
        extractError(error);
    }
};

/**
 * GET /dashboard/zones/{zoneId}/valves
 * Returns valves for a specific zone.
 *
 * @param {number} zoneId
 * @returns {Promise<ValveDTO[]>}
 */
export const getValvesByZone = async (zoneId) => {
    try {
        const res = await api.get(`${BASE}/zones/${zoneId}/valves`);
        return res.data;
    } catch (error) {
        extractError(error);
    }
};

/**
 * GET /dashboard/valves/open
 * Returns all valves that are currently OPEN (actively irrigating).
 *
 * @returns {Promise<ValveDTO[]>}
 */
export const getOpenValves = async () => {
    try {
        const res = await api.get(`${BASE}/valves/open`);
        return res.data;
    } catch (error) {
        extractError(error);
    }
};
